404 -  Update the route pattern for compendia file downloads

nozomione opened the pull request:
## Issue Number

Closes #404

## Purpose/Implementation Notes

I've created the dynamic route `compendia/[type]/download/[organism_name]`  to handles user-selected compendia computed file downloads and adjusted the codebase accordingly.

Changes include:

- Added the route: `compendia/[type]/download/[organism_name]`, with the user-selected compendia ID as a query
- Updated the following methods in `useCompendia`:
  - `goToDownloadPage` (formerly `navigateToFileDownload`)
  - `downloadCompendia`
 - Updated the compendia tracking methods in `gtag` 

**Additional update:**

- Added `CompendiaContext` and for `useCompendiaContext` (previously `useCompendia`)
- Added a new hook `useDownloadCompedium` for the compedium file download
- Utilized `getServerSideProps` for the dynamic routes (`[type]` and `[organism_name]`)

## Types of changes
- New feature (non-breaking change which adds functionality)
- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

<!-- Put an `x` in the boxes that apply. -->

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

nozomione commented:
I've applied the feedback and made the necessary changes. This PR is now ready for your review. Thank you, David!

davidsmejia commented on src/components/Compendia/DownloadBlockForm.js:
```diff
@@ -181,9 +168,9 @@ export const Download = ({ compendia }) => {
-                    selectedOrganism={
-                      selectedOrganism
-                        ? formatString(selectedOrganism.primary_organism_name)
+                    compendium={
+                      compendium
+                        ? formatString(compendium.primary_organism_name)
```
```suggestion
                    selected={compendium === option}
```

We should be able to just compare the references to the objects here.

davidsmejia commented on src/components/shared/PageTitle.js:
```diff
@@ -24,6 +24,15 @@ export const PageTitle = ({ title = '' }) => {
       pageTitle = `About`
       break
+    case /\/compendia/.test(path):
+      if (path.includes('normalized')) {
+        pageTitle = 'Normalized Compendia'
+      } else if (path.includes('rna-seq')) {
+        pageTitle = 'RNA-seq Sample Compendia'
+      } else {
+        pageTitle = 'Download Compendia -'
+      }
```
Seems like you need to check the presence of `download` first otherwise it will always be one of the other two.

davidsmejia commented on src/pages/compendia/[type]/download/[organism_name].js:
```diff
@@ -0,0 +1,126 @@
+
+  return {
+    props: {
+      compendium: compendium || {},
```
Is this necessary? Would passing a null value matter in this instance if we also pass `notFound`? Similar to the other compendia page?

davidsmejia commented on src/pages/compendia/[type].js:
```diff
@@ -1,69 +1,80 @@
-    fetchCompendia()
-  }, [])
+  ]
+  // sets the active tab index based on the selected type
```
```suggestion
```

davidsmejia commented on src/pages/compendia/[type].js:
```diff
@@ -1,69 +1,80 @@
-  }, [])
+  ]
+  // sets the active tab index based on the selected type
+  const activeIndex = tabConfigs.findIndex(({ id }) => id === type)
```
```suggestion
  const activeIndex = tabConfigs.findIndex((config) => config.type === type)
```

Maybe change id -> type and rewrite this way so you dont need the comment.

nozomione replied:
I've renamed it to `tab` to avoid ESLint `no-unused-vars` warning and removed the comment 👍

nozomione replied:
**UPDATE:**  I renamed `tabConfigs.tab` to `tabConfigs.type` and removed the destructuring when mapping the config in 691f095.

davidsmejia commented on src/contexts/CompendiaContext.js:
```diff
@@ -0,0 +1,24 @@
+
+export const CompendiaContext = createContext({})
+
+export const CompendiaContextProvider = ({ children }) => {
```
Why not instead of putting the `CompendiaContext.Provider` in _app you just use it normally and pass in the default values to the provider?

nozomione replied:
I've move `initialCompendia`, `initialType`, and the `useEffect` block from `useCompendiaContext` to `CompendiaContextProvider`,  and cleaned up the initialization in the `pages/compendia/[type]` component 👍

nozomione commented:
I've applied your feedback and this PR is ready for another look. Thank you David!

nozomione commented on src/components/Compendia/DownloadBlockForm.js:
```diff
@@ -1,34 +1,25 @@
 import { Row } from 'components/shared/Row'
 import { SearchBox } from 'components/shared/SearchBox'
 
-const DropDown = styled(Box)`
-  > div:nth-child(2) {
-    display: none;
-  }
-  &:focus-within > div:nth-child(2) {
-    display: block;
-    box-shadow: 0px 3px 4px rgba(0, 0, 0, 0.3);
-  }
-`
```
I've replaced this `DropDown` styled-componet  with `useRef` hook to handle the dropdown option toggling (in e248f1f).

nozomione commented:
Based on our 1:1 discussion, I've update the `filteredOptions` state to be a constant and updated the filtering logic with the `fuzzyFilterOnKey` helper. Additionally, I've removed the `DropDown` and adjusted the codebase accordingly. 

This PR is ready for your review. Thank you David!

davidsmejia reviewed (approved):
LGTM
